`crashdet_cancel()` doesnt cleanup all variables when USB printing

gudnimg opened the pull request:
This function doesn't look correct to me 🤔 I would think that this should behave similarly as when stopping the print via the LCD. This PR is my proposed improvement, un-tested at the moment.

Changed `UnconditionalStop()` to not close the SD card file if we're using Octoprint. Then there shouldnt be any file open.

Some of the variables which were not reset:
`isPrintPaused`
`pause_time`
`saved_start_position`
`saved_printing_type`

It seems like the bed heater could also be left enabled?



Fixes https://github.com/prusa3d/Prusa-Firmware/issues/4324

Change in memory:
Flash: -28 bytes
SRAM: 0 bytes

gudnimg committed:
crashdet_cancel doesnt cleanup all variables when using Octoprint

I would think that this should behave similarly as when stopping
the print via the LCD.

Changed UnconditionalSto()p to not close the SD card file
if we're using Octoprint. Then there shouldnt be any file open.

Some of the variables which were not reset:

isPrintPaused
pause_time
saved_start_position
saved_printing_type

Bed heater may be left on?

Change in memory:
Flash: -28 bytes
SRAM: 0 bytes

wavexx commented:
I generally agree here, I think this has always been overlooked recently

wavexx commented on Firmware/Marlin_main.cpp:
```diff
@@ -640,16 +640,9 @@ void crashdet_recover()
-	}
+/// Crash detection cancels the print
+void crashdet_cancel() {
+    print_stop();
```
The only difference I see here is that before we would send MSG_OCTOPRINT_CANCEL, but now we first call UnconditionalStop(), which clears card.sdprinting, so the condition in print_stop at ultralcd.cpp:5582 is always false...

this seems like a pre-existing bug?

gudnimg replied:
I just looked at it now and you're right!

![image](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/3db13374-5742-45f2-91a8-b3234b6b83d1)
![image](https://github.com/prusa3d/Prusa-Firmware/assets/8218499/0f826640-9d21-45b6-8964-c35595affabd)

gudnimg replied:
I added another commit to resolve this hopefully.

3d-gussner replied:
@gudnimg @wavexx is this solved? If so please use Resolve conversation to make it easier to merge.

wavexx reviewed (dismissed):
I agree here, I think this has always been overlooked

Maybe we can also call directly print_stop() when handling the CRASHDET_CANCEL command in Marlin_main.cpp:4035.

gudnimg committed:
Close SD file in print_stop instead of UnconditionalStop()

Change in memory:
Flash: -6 bytes
SRAM: 0 bytes

gudnimg committed:
Refactor code which resets crash detection setting

Add function crashdet_use_eeprom_setting

Change in memory:
Flash: -52 bytes
SRAM: 0 bytes

gudnimg commented:
I pushed another commit which optimizes setting the crash detection on or off.

gudnimg commented:
Adding 🚧 to the title as this PR is not tested

3d-gussner commented on Firmware/Marlin_main.cpp:
```diff
@@ -631,20 +631,17 @@ void crashdet_detected(uint8_t mask)
 {
-  if (!print_job_timer.isPaused()) crashdet_restore_print_and_continue();
-  if (lcd_crash_detect_enabled()) tmc2130_sg_stop_on_crash = true;
+	if (!print_job_timer.isPaused()) crashdet_restore_print_and_continue();
```
Change the line 620 `uint8_t yesno = lcd_show_yes_no_and_wait(false);` to `uint8_t yesno = lcd_show_yes_no_and_wait(false, LCD_LEFT_BUTTON_CHOICE);`
so RECOVER CRASH default is YES.
I think an accidental knob click should try to recover the crash instead of killing it.

3d-gussner reviewed (changes requested):
Tested on MK404
Started USB print
- [X] Triggered crash recovers
- [X] Triggered crash stop, stops the print correctly on printer and host
- [X] Crash det. [OFF] does work as it ignores the crashes

Approved only change I would like to see is that Recover crash default is YES
